Port `mkStdTx`

Anviking opened the pull request:
# Issue Number

#95 


# Overview

- [x] I ported `mkStdTx` from the legacy code-base.
- [x] The ported code allows the creation of witnesses from a transaction

# Comments

_Please, do not review yet, I think I'm splitting this up soon_

<!-- Additional comments or screenshots to attach if any -->
Just a draft. A few points:

- I probably added things in the wrong modules (easier for development, originally)
- I removed `SafeSigner`/`FakeSigner` abstraction 🤷‍♂️ 
- I removed shuffling of outputs
- Maybe some `EncryptedPassword` logic got lost
- While I don't believe I have broken something, it is not tested.
- Maybe some type-safety regarding the root secret key was lost — as in it might be easy to make a mistake and leak it.

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Rought scaffold (compiling with a few undefined)

Anviking committed:
even more scaffold

Anviking committed:
Remove usage of NE

Anviking committed:
rename

Anviking committed:
simplify by removing functions

Anviking committed:
even more?

Anviking committed:
MORE

Anviking committed:
refactor slightly

Anviking committed:
Relocate code to new module Signing

Anviking committed:
Remove SignTag

Anviking committed:
Replace TxAux with (Tx, [TxWitness])

Anviking committed:
protocolMagic unsafePerformIO

Anviking committed:
Replace (TxIn, TxOut)  with  (TxIn, Address)

KtorZ commented on src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -506,3 +519,25 @@ invariant
+protocolMagic :: Int32
+protocolMagic = fromMaybe testnet (envvar >>= readMaybe)
+  where
+    testnet = error "what /is/ the testnet pm?"
```
mainnet: 764824073
testnet: 1097911063

Thinking a second time aobut this, maybe we can actually default to `mainnet` :thinking: ? What do you think ?

Anviking replied:
🤔 Wouldn't it be "safest" to have no default? Or otherwise default to testnet? Such that `testnet` txs don't accidentally end up on mainnet?

Or would defaulting to `mainnet` give us fewer worries in production?

KtorZ replied:
Yeah right. Crashing with a clear message that we expect an ENV var with an appropriate protocol magic. I'd even go further and not accept _any_ value, but only the ones we recognize (adding staging: 633343913).

KtorZ commented on src/Cardano/Wallet/Primitive/Signing.hs:
```diff
@@ -0,0 +1,114 @@
+signTag :: ByteString
+signTag = signTxTag <> network
+  where
+    signTxTag = "\x01"
```
I'd still keep the sign tag as an ADT with a single option `SignTx` and write a proper encoder for that type :+1:

KtorZ commented on src/Cardano/Wallet/Primitive/Signing.hs:
```diff
@@ -0,0 +1,114 @@
+--
+-- TODO: re-add shuffle
+-- TODO: I removed FakeSigner/SafeSigner. Might be wrong.
+mkStdTx :: (Address -> Either e (Key 'RootK XPrv))
```
You'll also need the passphrase to make the signature here (because the `XPrv` is actually encrypted)

Anviking committed:
wip

Anviking committed:
test

Anviking committed:
test

Anviking committed:
test

Anviking committed:
Merge remote-tracking branch 'origin/master' into anviking/95/tx-witnesses

KtorZ commented on src/Cardano/Wallet.hs:
```diff
@@ -117,6 +132,16 @@ mkWalletLayer db network = WalletLayer
+        Just w -> do
+            let state = getState w
+            let rootXPrv = error "TODO"
+            let password = error "TODO"
```
Password would have to be passed as a parameter here, we can't look it up because we don't store it :+1:

KtorZ commented on src/Cardano/Wallet/Binary.hs:
```diff
@@ -426,7 +428,7 @@ decodeTxWitness = do
     t <- CBOR.decodeWord8
     case t of
-        0 -> CBOR.decodeTag *> (PublicKeyWitness <$> CBOR.decodeBytes)
+        0 -> CBOR.decodeTag *> (PublicKeyWitness <$> CBOR.decodeBytes <*> undefined)
```
What is the `undefined` for here :thinking:  ?

KtorZ commented on src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -330,6 +344,12 @@ instance Buildable Direction where
+data TxWitness
+    = PublicKeyWitness ByteString (Hash "tx")
+    | ScriptWitness ByteString
+    | RedeemWitness ByteString
```
We only _care_ about PublicKey witness, others are irrelevant to us. For now at least.

Anviking committed:
signTx

Anviking committed:
walletlayer

Anviking committed:
re-add SignTag and tx hashing

Anviking committed:
remove all SignTags but SignTx

Anviking committed:
'mainnet' and 'staging' build flags & default to testnet

This commit does not replace existing CLI logic for choosing network,
which would be a natural follow-up, if we were to go down this path.

Anviking committed:
some cleanup

Anviking committed:
Revert "'mainnet' and 'staging' build flags & default to testnet"

This reverts commit c5843e5ecb3fac94a643806d092a5a7a5251e72d.
